File: backend/utils/coordinate_mapping.py
"""
TimberGem Coordinate Mapping System

This module provides a centralized, clean coordinate transformation system for TimberGem.
PDF coordinates (in points) are the single source of truth, with clear transformations
to image coordinates (pixels) and canvas coordinates (UI pixels).

Key Principles:
- PDF coordinates: top-left origin, points (72 DPI) - PyMuPDF uses top-left, not bottom-left!
- Image coordinates: top-left origin, pixels at specific DPI
- Canvas coordinates: top-left origin, scaled pixels for UI display
"""

from dataclasses import dataclass, asdict
from typing import Dict, Tuple
import math


@dataclass
class PDFCoordinates:
    """
    PDF coordinates in points (72 DPI), origin at top-left.
    This is the single source of truth for all coordinate systems.
    Note: PyMuPDF uses top-left origin, not the PDF standard's bottom-left!
    """

    left: float
    top: float
    width: float
    height: float

    def to_dict(self) -> Dict[str, float]:
        return {
            "left_points": self.left,
            "top_points": self.top,
            "width_points": self.width,
            "height_points": self.height,
        }

    def to_rect_tuple(self) -> Tuple[float, float, float, float]:
        """Return as (x0, y0, x1, y1) tuple for PyMuPDF Rect"""
        return (self.left, self.top, self.left + self.width, self.top + self.height)


@dataclass
class ImageCoordinates:
    """
    Image coordinates in pixels at specific DPI, origin at top-left.
    Used for pixmap generation and image-based operations.
    """

    left: int
    top: int
    width: int
    height: int
    dpi: int

    def to_dict(self) -> Dict[str, int]:
        return {
            "left_pixels": self.left,
            "top_pixels": self.top,
            "width_pixels": self.width,
            "height_pixels": self.height,
            "dpi": self.dpi,
        }


@dataclass
class CanvasCoordinates:
    """
    Canvas coordinates in pixels for UI display, origin at top-left.
    Includes canvas dimensions for scale calculations.
    """

    left: float
    top: float
    width: float
    height: float
    canvas_width: float
    canvas_height: float

    def to_dict(self) -> Dict[str, float]:
        return {
            "left_canvas_pixels": self.left,
            "top_canvas_pixels": self.top,
            "width_canvas_pixels": self.width,
            "height_canvas_pixels": self.height,
            "canvas_width_pixels": self.canvas_width,
            "canvas_height_pixels": self.canvas_height,
        }


@dataclass
class ClippingCoordinates:
    """
    Coordinates within a legend clipping image (pixels at clipping DPI).
    Used for symbol annotations within extracted legend clippings.
    """

    left_pixels: int
    top_pixels: int
    width_pixels: int
    height_pixels: int
    clipping_dpi: int

    def to_dict(self) -> Dict[str, int]:
        return {
            "left_clipping_pixels": self.left_pixels,
            "top_clipping_pixels": self.top_pixels,
            "width_clipping_pixels": self.width_pixels,
            "height_clipping_pixels": self.height_pixels,
            "clipping_dpi": self.clipping_dpi,
        }


@dataclass
class PageMetadata:
    """
    Complete page metadata for coordinate transformations.
    Contains all information needed for accurate coordinate mapping.
    """

    page_number: int

    # PDF properties (points, 72 DPI)
    pdf_width_points: float
    pdf_height_points: float
    pdf_rotation_degrees: int

    # Standard image properties (for display)
    image_width_pixels: int
    image_height_pixels: int
    image_dpi: int

    # High-resolution image properties (for clipping extraction)
    high_res_image_width_pixels: int
    high_res_image_height_pixels: int
    high_res_dpi: int

    def to_dict(self) -> Dict:
        return asdict(self)

    @classmethod
    def from_dict(cls, data: Dict) -> "PageMetadata":
        return cls(**data)


class CoordinateTransformer:
    """
    Centralized coordinate transformation engine.
    Handles all transformations between PDF, image, and canvas coordinate systems.
    """

    def __init__(self, page_metadata: PageMetadata):
        self.page_metadata = page_metadata

        # Pre-calculate transformation factors for efficiency
        self.pdf_to_image_scale = self._calculate_pdf_to_image_scale()
        self.image_to_pdf_scale = 1.0 / self.pdf_to_image_scale

        self.pdf_to_high_res_scale = self._calculate_pdf_to_high_res_scale()
        self.high_res_to_pdf_scale = 1.0 / self.pdf_to_high_res_scale

    def _calculate_pdf_to_image_scale(self) -> float:
        """Calculate scale factor from PDF points to standard image pixels"""
        return self.page_metadata.image_dpi / 72.0  # 72 points per inch

    def _calculate_pdf_to_high_res_scale(self) -> float:
        """Calculate scale factor from PDF points to high-res image pixels"""
        return self.page_metadata.high_res_dpi / 72.0

    def pdf_to_image(self, pdf_coords: PDFCoordinates) -> ImageCoordinates:
        """
        Transform PDF coordinates to standard image coordinates.
        Both use top-left origin, so no Y-axis flip needed!
        """
        # Convert points to pixels - simple scaling, no coordinate system changes
        image_left = int(pdf_coords.left * self.pdf_to_image_scale)
        image_top = int(pdf_coords.top * self.pdf_to_image_scale)
        image_width = int(pdf_coords.width * self.pdf_to_image_scale)
        image_height = int(pdf_coords.height * self.pdf_to_image_scale)

        return ImageCoordinates(
            left=image_left,
            top=image_top,
            width=image_width,
            height=image_height,
            dpi=self.page_metadata.image_dpi,
        )

    def image_to_pdf(self, image_coords: ImageCoordinates) -> PDFCoordinates:
        """
        Transform image coordinates back to PDF coordinates.
        Both use top-left origin, so no Y-axis flip needed!
        """
        # Use the actual DPI of the image coordinates to scale back to points.
        # This ensures 300 DPI detection/image coords map correctly even if the
        # page's standard image_dpi differs.
        points_per_pixel = 72.0 / float(image_coords.dpi)

        pdf_left = image_coords.left * points_per_pixel
        pdf_top = image_coords.top * points_per_pixel
        pdf_width = image_coords.width * points_per_pixel
        pdf_height = image_coords.height * points_per_pixel

        return PDFCoordinates(
            left=pdf_left, top=pdf_top, width=pdf_width, height=pdf_height
        )

    def image_to_canvas(
        self, image_coords: ImageCoordinates, canvas_width: float, canvas_height: float
    ) -> CanvasCoordinates:
        """
        Transform image coordinates to canvas coordinates.
        Maintains aspect ratio using uniform scaling.
        """
        # Calculate scale factors
        scale_x = canvas_width / self.page_metadata.image_width_pixels
        scale_y = canvas_height / self.page_metadata.image_height_pixels

        # Use uniform scale to maintain aspect ratio
        scale = min(scale_x, scale_y)

        return CanvasCoordinates(
            left=image_coords.left * scale,
            top=image_coords.top * scale,
            width=image_coords.width * scale,
            height=image_coords.height * scale,
            canvas_width=canvas_width,
            canvas_height=canvas_height,
        )

    def canvas_to_image(self, canvas_coords: CanvasCoordinates) -> ImageCoordinates:
        """
        Transform canvas coordinates back to image coordinates.
        Reverses the uniform scaling applied in image_to_canvas.
        """
        # Calculate reverse scale factors
        scale_x = canvas_coords.canvas_width / self.page_metadata.image_width_pixels
        scale_y = canvas_coords.canvas_height / self.page_metadata.image_height_pixels

        # Use uniform scale (same as forward transformation)
        scale = min(scale_x, scale_y)

        return ImageCoordinates(
            left=int(canvas_coords.left / scale),
            top=int(canvas_coords.top / scale),
            width=int(canvas_coords.width / scale),
            height=int(canvas_coords.height / scale),
            dpi=self.page_metadata.image_dpi,
        )

    def canvas_to_pdf(self, canvas_coords: CanvasCoordinates) -> PDFCoordinates:
        """
        Transform canvas coordinates directly to PDF coordinates.
        Convenience method that combines canvas_to_image and image_to_pdf.
        """
        image_coords = self.canvas_to_image(canvas_coords)
        return self.image_to_pdf(image_coords)

    def pdf_to_canvas(
        self, pdf_coords: PDFCoordinates, canvas_width: float, canvas_height: float
    ) -> CanvasCoordinates:
        """
        Transform PDF coordinates directly to canvas coordinates.
        Convenience method that combines pdf_to_image and image_to_canvas.
        """
        image_coords = self.pdf_to_image(pdf_coords)
        return self.image_to_canvas(image_coords, canvas_width, canvas_height)

    def get_canvas_dimensions_for_aspect_ratio(
        self, max_width: float, max_height: float
    ) -> Tuple[float, float]:
        """
        Calculate optimal canvas dimensions that maintain image aspect ratio.
        Used by frontend to determine proper canvas sizing.
        """
        image_aspect_ratio = (
            self.page_metadata.image_width_pixels
            / self.page_metadata.image_height_pixels
        )
        max_aspect_ratio = max_width / max_height

        if image_aspect_ratio > max_aspect_ratio:
            # Image is wider - fit to width
            canvas_width = max_width
            canvas_height = max_width / image_aspect_ratio
        else:
            # Image is taller - fit to height
            canvas_height = max_height
            canvas_width = max_height * image_aspect_ratio

        return canvas_width, canvas_height


class ClippingCoordinateTransformer:
    """
    Handles coordinate transformations within legend clippings.
    Used for symbol annotations that occur within extracted legend images.
    """

    def __init__(
        self,
        legend_pdf_coords: PDFCoordinates,
        clipping_dpi: int,
        page_metadata: PageMetadata,
    ):
        self.legend_pdf_coords = legend_pdf_coords
        self.clipping_dpi = clipping_dpi
        self.page_metadata = page_metadata

        # Calculate clipping dimensions in pixels
        self.clipping_width_pixels = int(legend_pdf_coords.width * clipping_dpi / 72.0)
        self.clipping_height_pixels = int(
            legend_pdf_coords.height * clipping_dpi / 72.0
        )

        # Pre-calculate scale factors
        self.points_per_pixel = 72.0 / clipping_dpi

    def symbol_canvas_to_pdf(
        self, symbol_canvas_coords: CanvasCoordinates
    ) -> PDFCoordinates:
        """
        Transform symbol annotation coordinates to absolute PDF coordinates.
        This is the key method for mapping symbols back to the original PDF.
        """
        # Step 1: Canvas → Clipping image coordinates
        clipping_coords = self.canvas_to_clipping(symbol_canvas_coords)

        # Step 2: Clipping image coordinates → PDF coordinates
        return self.clipping_to_pdf(clipping_coords)

    def canvas_to_clipping(
        self, canvas_coords: CanvasCoordinates
    ) -> ClippingCoordinates:
        """
        Transform canvas coordinates to clipping image coordinates.
        """
        # Calculate canvas scale factor (maintaining aspect ratio)
        canvas_scale_x = canvas_coords.canvas_width / self.clipping_width_pixels
        canvas_scale_y = canvas_coords.canvas_height / self.clipping_height_pixels
        canvas_scale = min(canvas_scale_x, canvas_scale_y)

        # Transform to clipping coordinates
        clipping_left = int(canvas_coords.left / canvas_scale)
        clipping_top = int(canvas_coords.top / canvas_scale)
        clipping_width = int(canvas_coords.width / canvas_scale)
        clipping_height = int(canvas_coords.height / canvas_scale)

        return ClippingCoordinates(
            left_pixels=clipping_left,
            top_pixels=clipping_top,
            width_pixels=clipping_width,
            height_pixels=clipping_height,
            clipping_dpi=self.clipping_dpi,
        )

    def clipping_to_pdf(self, clipping_coords: ClippingCoordinates) -> PDFCoordinates:
        """
        Transform clipping coordinates to absolute PDF coordinates.
        """
        # Convert clipping pixels to points
        pdf_x_offset = clipping_coords.left_pixels * self.points_per_pixel
        pdf_y_offset = clipping_coords.top_pixels * self.points_per_pixel
        pdf_symbol_width = clipping_coords.width_pixels * self.points_per_pixel
        pdf_symbol_height = clipping_coords.height_pixels * self.points_per_pixel

        # Add legend offset to get absolute PDF coordinates
        # Note: Clipping coordinates have same origin as PDF (top-left of clipping = top-left of legend in PDF space)
        absolute_pdf_left = self.legend_pdf_coords.left + pdf_x_offset
        absolute_pdf_top = self.legend_pdf_coords.top + pdf_y_offset

        return PDFCoordinates(
            left=absolute_pdf_left,
            top=absolute_pdf_top,
            width=pdf_symbol_width,
            height=pdf_symbol_height,
        )

    def pdf_to_clipping(self, pdf_coords: PDFCoordinates) -> ClippingCoordinates:
        """
        Transform absolute PDF coordinates to clipping coordinates.
        Used for reverse transformations if needed.
        """
        # Calculate relative position within legend
        pdf_x_offset = pdf_coords.left - self.legend_pdf_coords.left
        pdf_y_offset = pdf_coords.top - self.legend_pdf_coords.top

        # Convert to clipping pixels
        clipping_left = int(pdf_x_offset / self.points_per_pixel)
        clipping_top = int(pdf_y_offset / self.points_per_pixel)
        clipping_width = int(pdf_coords.width / self.points_per_pixel)
        clipping_height = int(pdf_coords.height / self.points_per_pixel)

        return ClippingCoordinates(
            left_pixels=clipping_left,
            top_pixels=clipping_top,
            width_pixels=clipping_width,
            height_pixels=clipping_height,
            clipping_dpi=self.clipping_dpi,
        )


def validate_coordinates(coords, coord_type: str) -> bool:
    """
    Validate coordinate objects for basic sanity checks.
    """
    if coord_type == "pdf":
        return (
            coords.width > 0
            and coords.height > 0
            and coords.left >= 0
            and coords.top >= 0
        )
    elif coord_type == "image":
        return (
            coords.width > 0
            and coords.height > 0
            and coords.left >= 0
            and coords.top >= 0
            and coords.dpi > 0
        )
    elif coord_type == "canvas":
        return (
            coords.width > 0
            and coords.height > 0
            and coords.left >= 0
            and coords.top >= 0
            and coords.canvas_width > 0
            and coords.canvas_height > 0
        )
    elif coord_type == "clipping":
        return (
            coords.width_pixels > 0
            and coords.height_pixels > 0
            and coords.left_pixels >= 0
            and coords.top_pixels >= 0
            and coords.clipping_dpi > 0
        )

    return False


# Constants
DEFAULT_DPI = 300
HIGH_RES_DPI = 300
POINTS_PER_INCH = 72.0
MAX_CANVAS_WIDTH = 1200
MAX_CANVAS_HEIGHT = 900

File: frontend/src/utils/coordinateMapping.js
/**
 * TimberGem Frontend Coordinate Mapping System
 * 
 * JavaScript implementation that mirrors the backend coordinate transformation system.
 * Ensures consistent coordinate handling between frontend UI and backend processing.
 */

// Constants
export const DEFAULT_DPI = 200;
export const HIGH_RES_DPI = 300;
export const POINTS_PER_INCH = 72.0;
export const MAX_CANVAS_WIDTH = 1200;
export const MAX_CANVAS_HEIGHT = 900;

/**
 * PDF coordinates in points (72 DPI), origin at top-left.
 * This is the single source of truth for all coordinate systems.
 * Note: PyMuPDF uses top-left origin, not the PDF standard's bottom-left!
 */
export class PDFCoordinates {
    constructor(left, top, width, height) {
        this.left = left;
        this.top = top;
        this.width = width;
        this.height = height;
    }

    toDict() {
        return {
            left_points: this.left,
            top_points: this.top,
            width_points: this.width,
            height_points: this.height
        };
    }

    toRectTuple() {
        return [this.left, this.top, this.left + this.width, this.top + this.height];
    }
}

/**
 * Image coordinates in pixels at specific DPI, origin at top-left.
 * Used for pixmap generation and image-based operations.
 */
export class ImageCoordinates {
    constructor(left, top, width, height, dpi) {
        this.left = left;
        this.top = top;
        this.width = width;
        this.height = height;
        this.dpi = dpi;
    }

    toDict() {
        return {
            left_pixels: this.left,
            top_pixels: this.top,
            width_pixels: this.width,
            height_pixels: this.height,
            dpi: this.dpi
        };
    }
}

/**
 * Canvas coordinates in pixels for UI display, origin at top-left.
 * Includes canvas dimensions for scale calculations.
 */
export class CanvasCoordinates {
    constructor(left, top, width, height, canvasWidth, canvasHeight) {
        this.left = left;
        this.top = top;
        this.width = width;
        this.height = height;
        this.canvasWidth = canvasWidth;
        this.canvasHeight = canvasHeight;
    }

    toDict() {
        return {
            left_canvas_pixels: this.left,
            top_canvas_pixels: this.top,
            width_canvas_pixels: this.width,
            height_canvas_pixels: this.height,
            canvas_width_pixels: this.canvasWidth,
            canvas_height_pixels: this.canvasHeight
        };
    }
}

/**
 * Coordinates within a legend clipping image (pixels at clipping DPI).
 * Used for symbol annotations within extracted legend clippings.
 */
export class ClippingCoordinates {
    constructor(leftPixels, topPixels, widthPixels, heightPixels, clippingDpi) {
        this.leftPixels = leftPixels;
        this.topPixels = topPixels;
        this.widthPixels = widthPixels;
        this.heightPixels = heightPixels;
        this.clippingDpi = clippingDpi;
    }

    toDict() {
        return {
            left_clipping_pixels: this.leftPixels,
            top_clipping_pixels: this.topPixels,
            width_clipping_pixels: this.widthPixels,
            height_clipping_pixels: this.heightPixels,
            clipping_dpi: this.clippingDpi
        };
    }
}

/**
 * Complete page metadata for coordinate transformations.
 * Contains all information needed for accurate coordinate mapping.
 */
export class PageMetadata {
    constructor(data) {
        this.pageNumber = data.page_number;
        this.pdfWidthPoints = data.pdf_width_points;
        this.pdfHeightPoints = data.pdf_height_points;
        this.pdfRotationDegrees = data.pdf_rotation_degrees;
        this.imageWidthPixels = data.image_width_pixels;
        this.imageHeightPixels = data.image_height_pixels;
        this.imageDpi = data.image_dpi;
        this.highResImageWidthPixels = data.high_res_image_width_pixels;
        this.highResImageHeightPixels = data.high_res_image_height_pixels;
        this.highResDpi = data.high_res_dpi;
    }

    toDict() {
        return {
            page_number: this.pageNumber,
            pdf_width_points: this.pdfWidthPoints,
            pdf_height_points: this.pdfHeightPoints,
            pdf_rotation_degrees: this.pdfRotationDegrees,
            image_width_pixels: this.imageWidthPixels,
            image_height_pixels: this.imageHeightPixels,
            image_dpi: this.imageDpi,
            high_res_image_width_pixels: this.highResImageWidthPixels,
            high_res_image_height_pixels: this.highResImageHeightPixels,
            high_res_dpi: this.highResDpi
        };
    }
}

/**
 * Centralized coordinate transformation engine.
 * Handles all transformations between PDF, image, and canvas coordinate systems.
 */
export class CoordinateTransformer {
    constructor(pageMetadata) {
        this.pageMetadata = pageMetadata;
        
        // Pre-calculate transformation factors for efficiency
        this.pdfToImageScale = this._calculatePdfToImageScale();
        this.imageToPdfScale = 1.0 / this.pdfToImageScale;
        
        this.pdfToHighResScale = this._calculatePdfToHighResScale();
        this.highResToPdfScale = 1.0 / this.pdfToHighResScale;
    }

    _calculatePdfToImageScale() {
        return this.pageMetadata.imageDpi / 72.0; // 72 points per inch
    }

    _calculatePdfToHighResScale() {
        return this.pageMetadata.highResDpi / 72.0;
    }

    /**
     * Transform PDF coordinates to standard image coordinates.
     * Both use top-left origin, so no Y-axis flip needed!
     */
    pdfToImage(pdfCoords) {
        // Convert points to pixels - simple scaling, no coordinate system changes
        const imageLeft = Math.round(pdfCoords.left * this.pdfToImageScale);
        const imageTop = Math.round(pdfCoords.top * this.pdfToImageScale);
        const imageWidth = Math.round(pdfCoords.width * this.pdfToImageScale);
        const imageHeight = Math.round(pdfCoords.height * this.pdfToImageScale);
        
        return new ImageCoordinates(imageLeft, imageTop, imageWidth, imageHeight, this.pageMetadata.imageDpi);
    }

    /**
     * Transform image coordinates back to PDF coordinates.
     * Both use top-left origin, so no Y-axis flip needed!
     */
    imageToPdf(imageCoords) {
        // Convert pixels to points - simple scaling, no coordinate system changes
        const pdfLeft = imageCoords.left * this.imageToPdfScale;
        const pdfTop = imageCoords.top * this.imageToPdfScale;
        const pdfWidth = imageCoords.width * this.imageToPdfScale;
        const pdfHeight = imageCoords.height * this.imageToPdfScale;
        
        return new PDFCoordinates(pdfLeft, pdfTop, pdfWidth, pdfHeight);
    }

    /**
     * Transform image coordinates to canvas coordinates.
     * Maintains aspect ratio using uniform scaling.
     */
    imageToCanvas(imageCoords, canvasWidth, canvasHeight) {
        // Calculate scale factors
        const scaleX = canvasWidth / this.pageMetadata.imageWidthPixels;
        const scaleY = canvasHeight / this.pageMetadata.imageHeightPixels;
        
        // Use uniform scale to maintain aspect ratio
        const scale = Math.min(scaleX, scaleY);
        
        return new CanvasCoordinates(
            imageCoords.left * scale,
            imageCoords.top * scale,
            imageCoords.width * scale,
            imageCoords.height * scale,
            canvasWidth,
            canvasHeight
        );
    }

    /**
     * Transform canvas coordinates back to image coordinates.
     * Reverses the uniform scaling applied in imageToCanvas.
     */
    canvasToImage(canvasCoords) {
        // Calculate reverse scale factors
        const scaleX = canvasCoords.canvasWidth / this.pageMetadata.imageWidthPixels;
        const scaleY = canvasCoords.canvasHeight / this.pageMetadata.imageHeightPixels;
        
        // Use uniform scale (same as forward transformation)
        const scale = Math.min(scaleX, scaleY);
        
        return new ImageCoordinates(
            Math.round(canvasCoords.left / scale),
            Math.round(canvasCoords.top / scale),
            Math.round(canvasCoords.width / scale),
            Math.round(canvasCoords.height / scale),
            this.pageMetadata.imageDpi
        );
    }

    /**
     * Transform canvas coordinates directly to PDF coordinates.
     * Convenience method that combines canvasToImage and imageToPdf.
     */
    canvasToPdf(canvasCoords) {
        const imageCoords = this.canvasToImage(canvasCoords);
        return this.imageToPdf(imageCoords);
    }

    /**
     * Transform PDF coordinates directly to canvas coordinates.
     * Convenience method that combines pdfToImage and imageToCanvas.
     */
    pdfToCanvas(pdfCoords, canvasWidth, canvasHeight) {
        const imageCoords = this.pdfToImage(pdfCoords);
        return this.imageToCanvas(imageCoords, canvasWidth, canvasHeight);
    }

    /**
     * Calculate optimal canvas dimensions that maintain image aspect ratio.
     * Used by frontend to determine proper canvas sizing.
     */
    getCanvasDimensionsForAspectRatio(maxWidth, maxHeight) {
        const imageAspectRatio = this.pageMetadata.imageWidthPixels / this.pageMetadata.imageHeightPixels;
        const maxAspectRatio = maxWidth / maxHeight;
        
        let canvasWidth, canvasHeight;
        
        if (imageAspectRatio > maxAspectRatio) {
            // Image is wider - fit to width
            canvasWidth = maxWidth;
            canvasHeight = maxWidth / imageAspectRatio;
        } else {
            // Image is taller - fit to height
            canvasHeight = maxHeight;
            canvasWidth = maxHeight * imageAspectRatio;
        }
        
        return { width: canvasWidth, height: canvasHeight };
    }
}

/**
 * Handles coordinate transformations within legend clippings.
 * Used for symbol annotations that occur within extracted legend images.
 */
export class ClippingCoordinateTransformer {
    constructor(legendPdfCoords, clippingDpi, pageMetadata) {
        this.legendPdfCoords = legendPdfCoords;
        this.clippingDpi = clippingDpi;
        this.pageMetadata = pageMetadata;
        
        // Calculate clipping dimensions in pixels
        this.clippingWidthPixels = Math.round(legendPdfCoords.width * clippingDpi / 72.0);
        this.clippingHeightPixels = Math.round(legendPdfCoords.height * clippingDpi / 72.0);
        
        // Pre-calculate scale factors
        this.pointsPerPixel = 72.0 / clippingDpi;
    }

    /**
     * Transform symbol annotation coordinates to absolute PDF coordinates.
     * This is the key method for mapping symbols back to the original PDF.
     */
    symbolCanvasToPdf(symbolCanvasCoords) {
        // Step 1: Canvas → Clipping image coordinates
        const clippingCoords = this.canvasToClipping(symbolCanvasCoords);
        
        // Step 2: Clipping image coordinates → PDF coordinates
        return this.clippingToPdf(clippingCoords);
    }

    /**
     * Transform canvas coordinates to clipping image coordinates.
     */
    canvasToClipping(canvasCoords) {
        // Calculate canvas scale factor (maintaining aspect ratio)
        const canvasScaleX = canvasCoords.canvasWidth / this.clippingWidthPixels;
        const canvasScaleY = canvasCoords.canvasHeight / this.clippingHeightPixels;
        const canvasScale = Math.min(canvasScaleX, canvasScaleY);
        
        // Transform to clipping coordinates
        const clippingLeft = Math.round(canvasCoords.left / canvasScale);
        const clippingTop = Math.round(canvasCoords.top / canvasScale);
        const clippingWidth = Math.round(canvasCoords.width / canvasScale);
        const clippingHeight = Math.round(canvasCoords.height / canvasScale);
        
        return new ClippingCoordinates(
            clippingLeft, clippingTop, clippingWidth, clippingHeight, this.clippingDpi
        );
    }

    /**
     * Transform clipping coordinates to absolute PDF coordinates.
     */
    clippingToPdf(clippingCoords) {
        // Convert clipping pixels to points
        const pdfXOffset = clippingCoords.leftPixels * this.pointsPerPixel;
        const pdfYOffset = clippingCoords.topPixels * this.pointsPerPixel;
        const pdfSymbolWidth = clippingCoords.widthPixels * this.pointsPerPixel;
        const pdfSymbolHeight = clippingCoords.heightPixels * this.pointsPerPixel;
        
        // Add legend offset to get absolute PDF coordinates
        // Note: Clipping coordinates have same origin as PDF (top-left of clipping = top-left of legend in PDF space)
        const absolutePdfLeft = this.legendPdfCoords.left + pdfXOffset;
        const absolutePdfTop = this.legendPdfCoords.top + pdfYOffset;
        
        return new PDFCoordinates(absolutePdfLeft, absolutePdfTop, pdfSymbolWidth, pdfSymbolHeight);
    }

    /**
     * Transform absolute PDF coordinates to clipping coordinates.
     * Used for reverse transformations if needed.
     */
    pdfToClipping(pdfCoords) {
        // Calculate relative position within legend
        const pdfXOffset = pdfCoords.left - this.legendPdfCoords.left;
        const pdfYOffset = pdfCoords.top - this.legendPdfCoords.top;
        
        // Convert to clipping pixels
        const clippingLeft = Math.round(pdfXOffset / this.pointsPerPixel);
        const clippingTop = Math.round(pdfYOffset / this.pointsPerPixel);
        const clippingWidth = Math.round(pdfCoords.width / this.pointsPerPixel);
        const clippingHeight = Math.round(pdfCoords.height / this.pointsPerPixel);
        
        return new ClippingCoordinates(
            clippingLeft, clippingTop, clippingWidth, clippingHeight, this.clippingDpi
        );
    }
}

/**
 * Validate coordinate objects for basic sanity checks.
 */
export function validateCoordinates(coords, coordType) {
    if (coordType === "pdf") {
        return coords.width > 0 && coords.height > 0 && coords.left >= 0 && coords.top >= 0;
    } else if (coordType === "image") {
        return coords.width > 0 && coords.height > 0 && coords.left >= 0 && coords.top >= 0 && coords.dpi > 0;
    } else if (coordType === "canvas") {
        return coords.width > 0 && coords.height > 0 && coords.left >= 0 && coords.top >= 0 && 
               coords.canvasWidth > 0 && coords.canvasHeight > 0;
    } else if (coordType === "clipping") {
        return coords.widthPixels > 0 && coords.heightPixels > 0 && coords.leftPixels >= 0 && 
               coords.topPixels >= 0 && coords.clippingDpi > 0;
    }
    
    return false;
}

/**
 * Utility function to create PageMetadata from backend response
 */
export function createPageMetadata(backendData) {
    return new PageMetadata(backendData);
}

/**
 * Utility function to get canvas dimensions for a page
 */
export function getOptimalCanvasDimensions(pageMetadata, maxWidth = MAX_CANVAS_WIDTH, maxHeight = MAX_CANVAS_HEIGHT) {
    const transformer = new CoordinateTransformer(pageMetadata);
    return transformer.getCanvasDimensionsForAspectRatio(maxWidth, maxHeight);
} 